Trait support

dshoreman opened the pull request:
Adds underlying Trait prototypes, builders etc required for phpactor/phpactor#726.

#### Todo
- [x] Make it work
- [x] Create new `ClassLikeUpdater` abstract class
- [x] Move common `TraitUpdater` and `ClassUpdater` code to `ClassLikeUpdater`
- [x] Refactor out switch in `updateProperties`  method
- [x] Commit patch to phpactor/code-transform
- [x] Submit PR for phpactor/code-transform patch
- [x] Tests for `Adapter\WorseReflection\WorseBuilderFactoryTest`
- [x] Tests for `Adapter\UpdaterTestCase`

#### Notes
* Some places had type hints and return types loosened from `ClassFoo` to `ClassLikeFoo` in order to receive `Trait`-related objects. Might need to rework this to avoid accepting `Interface`-related things where we don't want them.
* `ClassBuilder` and `TraitBuilder`, along with `ClassUpdater` and `TraitUpdater` share many similarities. Could probably refactor these too to avoid too much duplication potentially causing issues down the line.

dantleech commented:
> Some places had type hints and return types loosened from ClassFoo to ClassLikeFoo in order to receive Trait-related objects. Might need to rework this to avoid accepting Interface-related things where we don't want them.

This is probably acceptable, in cases where we know we return a more specific type we can add a virtual method declaration (e.g. `@method Acme\ActualImplementation foobar()`

> ClassBuilder and TraitBuilder, along with ClassUpdater and TraitUpdater share many similarities. Could probably refactor these too to avoid too much duplication potentially causing issues down the line.

For sure, although the builders don't worry as much as the updaters. The updating code can be broadly refactored, as currently it is very loose and clumsy.

dshoreman commented:
> Looks good so far :)

Thanks! That's a welcome and much need sanity check so I know I'm not going down a completely wrong path ;)

> Perhaps a first step woul be to create an abstract `ClassLikeUpdater` and push some methods up?

That's pretty much exactly what I was thinking, but I didn't want to go ahead if you preferred a simple copy approach. I'll work on this later today and see what I can come up with

dshoreman commented:
First stage of the `ClassLikeUpdater` is now pushed up, but `updateProperties()` and `updateConstants()` share a lot of code, so it's not quite finished yet.

Attempts at moving the first `foreach` loop without the rest of the method proved troublesome - the exact same code that passed tests in the "real" classes would fail when cut/pasted into the abstract. I've worked around it for now by moving the entire method and adding a `switch` where the code differed. It's not the most beautiful code in the world, but it's something I can work on before this gets merged.

To avoid any issues, I first want to add some tests to `Adapter\UpdaterTestCase`. Once that's done I'll come back and do some more refactoring within `ClassLikeUpdater`... hopefully including those ugly `switch` statements.

dantleech commented on lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php:
```diff
@@ -0,0 +1,168 @@
+                break;
+            case 'TraitDeclaration':
+                $lastConstant = $classNode->traitMembers->openBrace;
+                $memberDeclarations = $classNode->traitMembers->traitMemberDeclarations;
```
I guess maybe we can just add an abstract method `memberDeclarations(): array`? which returns `Node[]` ?

dshoreman replied:
That's the most likely option, along with something similar for the `$lastConstant` definition.

That said, I completely forgot traits don't actually support constants. With that in mind, I'll probably move `updateConstants()` back to the concrete `ClassUpdater` later today.

Hopefully nothing breaks in the process, but it will at least cut out one of the two switches.

dantleech commented:
:+1: looks good so far :)

dantleech commented on lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php:
```diff
@@ -47,27 +47,18 @@ protected function resolvePropertyName(Node $property)
+
+    abstract protected function memberDeclarations(Node $node): array;
+
     protected function updateProperties(Edits $edits, ClassLikePrototype $classPrototype, StatementNode $classNode)
```
No objection :+1:

dshoreman commented:
@dantleech All done! I've submitted phpactor/code-transform#23 which connects the dots if you want to try it locally. Otherwise, we should pretty much be good to go. Let me know if there's anything I've missed!

dantleech commented on tests/Adapter/WorseReflection/WorseBuilderFactoryTest.php:
```diff
@@ -145,18 +202,72 @@ public function testStaticMethod()
+    {
+        $source = $this->build('<?php trait Foobar { public function method($param = "1234") {} }');
+        $this->assertEquals('1234', (string) $source->traits()->first()->methods()->first()->parameters()->first()->defaultValue()->value());
+    }
```
Are these property / method tests necessary? These things are already tested individually, I guess the Trait logic doesn't affect them, in which case we don't benefit from duplicating the tests?

dshoreman replied:
It's possible, but honestly I'm not certain: I'm still very new to testing overall and before this PR I hadn't really dug into the internals of Phpactor at all. It wasn't until I added these extra tests that I found the commented `todo` block, so they definitely do _something_ - it's just a question of which ones are really necessary.

One thing I noticed with the property tests, the `...Typed` ones have a default value in the respective class versions. This had to be removed in the Trait copies to get them passing, but I put that down to 'skipping due to lack of default property value support](https://github.com/phpactor/code-builder/pull/9/files#diff-5ef7cbf371f45e47f8cab19dd507146dR114). Again I don't know enough to say for sure, but there certainly seems to be something different.

I guess they can't hurt as a "fireguard" in case the property/method handling ever changes? I'm happy to remove some though if duplication is still a concern.

dantleech commented:
Looks perfect :) Just left a comment about potentially duplicated tests

dshoreman commented:
After replying to that comment, I took another look at the tests. I've gone ahead and removed most of the duplicates, leaving behind:

* `testSimpleTrait`
* `testSimpleTraitWithNamespace`
* `testTraitWithProperty`
* `testTraitWithMethod`

...which should hopefully keep all bases covered.

I've also updated phpactor/code-transform#23 to get rid of the variable method. Let me know if you would rather an if/else instead of the shorthand and I'll update it.
